cmd/kubelet/app/server.go

```diff
@@ -385,36 +303,78 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
 
 	// Register current configuration with /configz endpoint
 	cfgz, cfgzErr := initConfigz(&s.KubeletConfiguration)
-	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicKubeletConfig) {
-		// Look for config on the API server. If it exists, replace s.KubeletConfiguration
-		// with it and continue. initKubeletConfigSync also starts the background thread that checks for new config.
-
-		// Don't do dynamic Kubelet configuration in runonce mode
-		if s.RunOnce == false {
-			remoteKC, err := initKubeletConfigSync(s)
-			if err == nil {
-				// Update s (KubeletServer) with new config from API server
-				s.KubeletConfiguration = *remoteKC
-				// Ensure that /configz is up to date with the new config
-				if cfgzErr != nil {
-					glog.Errorf("was unable to register configz before due to %s, will not be able to set now", cfgzErr)
-				} else {
-					setConfigz(cfgz, &s.KubeletConfiguration)
-				}
-				// Update feature gates from the new config
-				err = utilfeature.DefaultFeatureGate.Set(s.KubeletConfiguration.FeatureGates)
-				if err != nil {
-					return err
-				}
-			} else {
-				glog.Errorf("failed to init dynamic Kubelet configuration sync: %v", err)
+
+	// launch nodeconfig controller if the dynamic config feature gate is enabled
+	dynamicConfigEnabled := utilfeature.DefaultFeatureGate.Enabled(features.DynamicKubeletConfig)
+	if dynamicConfigEnabled {
+		// `dynamicConfigEnabled` indicates the feature gate, which is presently used to gate the whole controller
+		// eventually, this restriction will be reduced to only gating the "dynamic" part of the controller,
+		// which is why `dynamicConfigEnabled` is still included as part of the `useDynamicConfig` predicate.
+		useDynamicConfig := dynamicConfigEnabled && !standaloneMode && !s.RunOnce
+
+		flagCfg := &componentconfigv1alpha1.KubeletConfiguration{}
+		err = api.Scheme.Convert(&s.KubeletConfiguration, flagCfg, nil)
+		if err != nil {
+			return err
+		}
+
+		// the NodeConfigDir must be specified to use the controller
+		if len(s.NodeConfigDir) == 0 {
+			return fmt.Errorf("--node-config-dir must be specified to use the node config controller")
+		}
+		// compute absolute path for NodeConfigDir, if necessary
+		var nodeConfigDir string
+		if s.NodeConfigDir[0] != '/' {
+			nodeConfigDir = filepath.Join(s.RootDirectory, s.NodeConfigDir)
+		} else {
+			nodeConfigDir = s.NodeConfigDir
+		}
+
+		var ncc *nodeconfig.NodeConfigController
+		if useDynamicConfig {
+			client, err := getKubeClient(s)
```

Comment:
Ok those fields are yanked, no cert regen.
Clients are regenerated from latest config so that the QPS, etc. parameters are up-to-date, and then the latest client is injected when starting the sync loop.